### PR TITLE
[ddp] allreduce grouping

### PR DESCRIPTION
**Pull request by daquexian**
类似 pytorch，让 parameter 的 .grad 分布在连续的显存（bucket）上，这样可以一次 allreduce 多个 parameter.grad。从 https://github.com/Oneflow-Inc/oneflow/pull/6854 中搬运了一些需要的 view 操作。

在 12 上跑 10 次 bs 为 1 的测试的结果是：
原 fuse_instruction 分支相对 pytorch 的平均加速比是 1.305
fuse_instruction 合并了本分支之后相对 pytorch 的平均加速比是 1.364

### CODE REVIEW

File: oneflow/core/functional/impl/array_functor.cpp [hunk 1]
```diff
@@ -1749,15 +1755,16 @@ class TensorGetItemFunctor {
...
+    // bool is_identity = [&]() {
+    //   if (target_shape.NumAxes() == 0) { return false; }
+    //   for (int i = 0; i < ndims; ++i) {
+    //     if (start[i] != 0 || end[i] != expand_input->shape()->At(i) || step[i] != 1) {
+    //       return false;
+    //     }
+    //   }
+    //   return true;
+    // }();
+    const bool is_identity = false;
```
Comment (daquexian) [inline oneflow/core/functional/impl/array_functor.cpp:1767]: 这里暂时用这种方式强行让 `x[:]` 也走 view （pytorch 也是这样的行为）

File: oneflow/core/autograd/autograd_engine.cpp [hunk 2]
```diff
@@ -55,12 +55,14 @@ Maybe<void> CopyOrAccGrad(AutogradMeta* autograd_meta, bool autograd_mode) {
...
     // for dy if dx is shared with dz.
+    // FIXME(jianhao): add a flag to control inplace accumulate grad.
     const auto& output = JUST(
-        functional::Add(autograd_meta->acc_grad(), current_grad, /*alpha=*/1, /*inplace=*/false));
+        functional::Add(autograd_meta->acc_grad(), current_grad, /*alpha=*/1, /*inplace=*/true));
     JUST(autograd_meta->set_acc_grad(output));
   } else {
     JUST(autograd_meta->set_acc_grad(current_grad));
   }
+  for (const auto& hook : autograd_meta->post_hooks()) { hook(autograd_meta->acc_grad()); }
```
Comment (daquexian) [inline oneflow/core/autograd/autograd_engine.cpp:65]: 这里添加了一个 post hook 的概念（不暴露给用户），是在完成 tensor.grad 的累加之后，以 tensor.grad 为参数执行的 hook，allreduce 以这种形式执行，pytorch 也是这样做的

File: oneflow/core/autograd/autograd_engine.cpp [hunk 3]
```diff
@@ -55,12 +55,14 @@ Maybe<void> CopyOrAccGrad(AutogradMeta* autograd_meta, bool autograd_mode) {
     //
     // As we know that dx = dz + dp / z and dy = dz, so it will lead to wrong value
     // for dy if dx is shared with dz.
+    // FIXME(jianhao): add a flag to control inplace accumulate grad.
     const auto& output = JUST(
-        functional::Add(autograd_meta->acc_grad(), current_grad, /*alpha=*/1, /*inplace=*/false));
+        functional::Add(autograd_meta->acc_grad(), current_grad, /*alpha=*/1, /*inplace=*/true));
```
Comment (daquexian) [inline oneflow/core/autograd/autograd_engine.cpp:60]: 在本分支的 ddp 下 inplace 为 true 是安全的，但在非 ddp 下是不安全的，后续需要想个办法

File: oneflow/core/functional/impl/array_functor.cpp [hunk 4]
```diff
@@ -1004,6 +1004,12 @@ class SliceBaseFunctor {
   Maybe<Tensor> operator()(const std::shared_ptr<one::Tensor>& x, const std::vector<int64_t>& start,
                            const std::vector<int64_t>& stop,
                            const std::vector<int64_t>& step) const {
+    if (x->is_local() && !(LazyMode::is_enabled())) {
+      // TODO: view support 0-dim tensor
+      if (!(x->shape()->NumAxes() < 1 || x->shape()->elem_cnt() < 1)) {
```
Comment (daquexian) [inline oneflow/core/functional/impl/array_functor.cpp:1009]: 在原本的 view 分支里这里两个 < 1 都是 <= 1

**Comment by daquexian** [comment]
@ouyangyu 可以帮忙测下 resnet50 ddp 在这个分支上的正确性吗

**Comment by ouyangyu** [comment]
resnet50 ddp  完整训练，没有问题。

https://oneflow-test.oss-cn-beijing.aliyuncs.com/OneBrain/commit/ddp_allreduce_grouping/20220124/4e16e7f3c/ResNet50-ddp/train/1n8g/resnet50_ddp_DCcpu_FP32_b192_1n8g_ddp_allreduce_grouping_20220124_101433330235093.log

File: oneflow/core/eager/blob_instruction_type.cpp [hunk 5]
```diff
@@ -129,7 +129,7 @@ void TensorViewInstructionType::Compute(vm::Instruction* instruction) const {
   OfBlob input_ofblob(device_ctx->stream(), ptr->eager_blob_object()->mut_blob());
   OfBlob view_ofblob(device_ctx->stream(), ptr->view_eager_blob_object()->mut_blob());
 
-  void* input_ptr = input_ofblob.mut_blob()->mut_dptr();
+  void* input_ptr = input_ofblob.mut_blob()->mut_raw_dptr();
```
Comment (daquexian) [inline oneflow/core/eager/blob_instruction_type.cpp:132]: 这个来自 view 分支

File: oneflow/core/functional/impl/array_functor.cpp [hunk 6]
```diff
@@ -1777,21 +1783,8 @@ class TensorGetItemFunctor {
...
-        }
-      }
-      return true;
-    }();
-    std::shared_ptr<one::Tensor> result;
-    if (is_identity) {
-      result = expand_input;
-    } else {
-      result = JUST(Slice(expand_input, start, end, step));
-    }
```
Comment (daquexian) [inline oneflow/core/functional/impl/array_functor.cpp:1794]: 这里是我对 view 代码的修改，在 slice 实际是 identity 的情况下（y = x[:]）也走 view 逻辑，和 pytorch 一致

File: oneflow/core/framework/tensor_methods.cpp [hunk 7]
```diff
@@ -132,6 +137,61 @@ Maybe<Tensor> Reshape(const std::shared_ptr<Tensor>& input, const Shape& shape)
   return output;
 }
 
+Maybe<Tensor> Slice(const std::shared_ptr<Tensor>& input, const std::vector<int64_t>& starts,
+                    const std::vector<int64_t>& ends, const std::vector<int64_t>& steps) {
+  if (!(input->is_eager() && input->is_local())) {
```
Comment (lixinqi) [inline oneflow/core/framework/tensor_methods.cpp:142]: CHECK_OR_RETURN

Comment (lixinqi) [reply oneflow/core/framework/tensor_methods.cpp:142]: 下同。

File: oneflow/api/python/framework/tensor.cpp [hunk 8]
```diff
@@ -175,6 +174,10 @@ ONEFLOW_API_PYBIND11_MODULE("", m) {
               throw std::runtime_error("You can only change gradient of leaf tensors.");
             }
           })
+      .def_property(
+          "_is_grad_acc_inplace",
```
Comment (lixinqi) [inline oneflow/api/python/framework/tensor.cpp:178]: 这个字段是和Pytorch对齐的吗？

File: oneflow/core/framework/tensor_impl.cpp [hunk 9]
```diff
@@ -57,7 +57,10 @@ Maybe<TensorArg> TensorImpl::current_grad() const {
 }
 
 Maybe<void> TensorImpl::set_acc_grad(const std::shared_ptr<Tensor>& grad) {
-  CHECK_NOTNULL_OR_RETURN(autograd_meta_);
+  if (!has_autograd_meta()) {
+    const auto& autograd_meta = NewAutogradMeta(requires_grad(), is_leaf());
+    this->set_autograd_meta(autograd_meta);
+  }
```
Comment (lixinqi) [inline oneflow/core/framework/tensor_impl.cpp:63]: 这个逻辑是不是应该封装在mutable_autograd_meta()里，本函数只做mutable_autograd_meta()->set_acc_grad()。因为本函数的名字没有义务处理此处新增的逻辑。

Comment (lixinqi) [reply oneflow/core/framework/tensor_impl.cpp:63]: 另外，不是特别明白为什么会发生都要set_acc_grad了，连autograd_meta都没有的情况。需要代码注释。

File: oneflow/core/functional/impl/array_functor.cpp (hunk 4 above)
Comment (lixinqi) [reply oneflow/core/functional/impl/array_functor.cpp:1037]: < 1是不是换成==0更直白？最好同时加代码注释。

File: oneflow/core/framework/tensor_impl.cpp (hunk 9 above)
Comment (daquexian) [reply oneflow/core/framework/tensor_impl.cpp:63]: > 这个逻辑是不是应该封装在mutable_autograd_meta()里

感觉可以

Comment (daquexian) [reply oneflow/core/framework/tensor_impl.cpp]: > 不是特别明白为什么会发生都要set_acc_grad了，连autograd_meta都没有的情况。需要代码注释。

确实，是最初的一版实现需要这个功能（在 loss.backward 之前设置 param 的 .grad），现在已经不需要了，我把它去掉了

File: oneflow/core/functional/impl/array_functor.cpp (hunk 4 above)
Comment (daquexian) [reply oneflow/core/functional/impl/array_functor.cpp]: 已换并更新注释。不过这里是不是应该 @Flowingsun007 确认一下 view 实际是不是可以在 0-d 和 0-size tensor 上工作，当时绕过它们应该只是一种随手的、防御性的做法？如果其实可以工作，这里的判断就可以去掉了，也和 pytorch 对齐了

Comment (Flowingsun007) [reply oneflow/core/functional/impl/array_functor.cpp]: 确实，当时加判断是一种防御性的做法，避免在0-d tensor下view可能会有问题；按道理放开也可以正常工作，后面需要试一下；0-size tensor那块需要支持view吗？（感觉是不需要的？这块需要调研下）

**Review by lixinqi — APPROVED** [review]

File: oneflow/core/functional/impl/array_functor.cpp (hunk 6 above)
Comment (Flowingsun007) [reply oneflow/core/functional/impl/array_functor.cpp:1794]: 代码里可以加一下note（怕后面忘了

File: oneflow/api/python/framework/tensor.cpp (hunk 8 above)
Comment (daquexian) [reply oneflow/api/python/framework/tensor.cpp:178]: 不是，这个是不暴露给用户的私有方法

**Review by Flowingsun007 — APPROVED** [review]

**Review by hjchen2 — APPROVED** [review]
